File: 20221114_Monday/20221114.rb
# Leetcode: 947. Most Stones Removed with Same Row or Column.
# https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 180 ms, faster than 100.00% of Ruby online submissions for Most Stones Removed with Same Row or Column.
# Memory Usage: 211.9 MB, less than 81.82% of Ruby online submissions for Most Stones Removed with Same Row or Column.
# @param {Integer[][]} stones
# @return {Integer}
# Based on:
# https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/discuss/197668/Count-the-Number-of-Islands-O(N)
def remove_stones(stones)
  # Based on:
  # https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/discuss/197668/Count-the-Number-of-Islands-O(N)
  # https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/discuss/198141/Python-concise-DFS
  # https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/discuss/2812465/It's-Literally-a-Graph%3A-DFS-and-Union-Find
  @islands = 0
  @f = {}
  stones.each do |i,j|
    uni(i,~j)
  end
  stones.size - @islands 
end

def find(x)
  if @f[x].nil?
    @f[x] = x
    @islands += 1
  end
  if x != @f[x]
    @f[x] = find(@f[x])
  end
  return @f[x]
end

def uni(x,y)
  x = find(x)
  y = find(y)
  if x != y
    @f[x] = y
    @islands -= 1
  end
end

File: 20221114_Monday/20221114.go
// Leetcode: 947. Most Stones Removed with Same Row or Column.
// https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 43 ms, faster than 65.08% of Go online submissions for Most Stones Removed with Same Row or Column.
// Memory Usage: 7.3 MB, less than 46.03% of Go online submissions for Most Stones Removed with Same Row or Column.
// 2022.11.14 Daily Challenge.
var isl int // islands.
var f map[int]int // For union find.
func removeStones(stones [][]int) int {
  // Based on:
  // https://leetcode.com/problems/most-stones-removed-with-same-row-or-column/discuss/197668/Count-the-Number-of-Islands-O(N)
  isl = 0
  f = make(map[int]int)
  for _, el := range stones {
    x,y := el[0],el[1]
    uni(x,-(1+y))
  }
  return len(stones) - isl
}

func find(x int) int {
  if _, ok := f[x]; ! ok {
    f[x] = x
    isl += 1
  }
  if x != f[x] {
    f[x] = find(f[x])
  }
  return f[x]
}

func uni(x int, y int) {
  x = find(x)
  y = find(y)
  if x != y {
    f[x] = y
    isl -= 1
  }
}

